src/mon/MgrMonitor.cc

```diff
@@ -987,6 +987,29 @@ int MgrMonitor::dump_metadata(const string& name, Formatter *f, ostream *err)
   return 0;
 }
 
+void MgrMonitor::print_nodes(Formatter *f)
+{
+  assert(f);
+
+  std::map<string, list<int> > mgrs;    // hostname => mgr
+  std::set<string> ls = map.get_all_names();
+  for (auto& name : ls) {
+    std::map<string,string> meta;
+    if (load_metadata(name, meta, NULL)) {
+      continue;
+    }
+   std::map<string, string>::iterator hostname = meta.find("hostname");
+    if (hostname == meta.end()) {
+      // not likely though
+      continue;
+    }
+    // FIXME: not sure any mgrmap_id valid here?
+    mgrs[hostname->second].push_back(0);
```

Comment:
could look up the gid by the "`name`" here. i.e.
```c++
uint64_t gid = 0;
if (map.get_active_name() == name) {
  gid = map.get_active_gid();
else {
  for (auto& standby : map.standbys) {
    if (standby.name == name) {
      gid = standby.gid;
      break;
    }
  }
}
mgrs[hostname->second].push_back(gid);
```

or, could add a utility method to `MgrMap` to do the lookup.

Reply:
I feel like I might be missing something about the intended purposes of this output.

Looking at the existing code for MDS and mon daemons, we're outputting their ranks (rather than their daemon IDs).  I'm not sure what the motivation was for that: knowing the rank is generally much less useful than the daemon name if you're going to e.g. go look on the node for logs etc.

Similarly, I'm not sure what the utility of knowing the gid is for a mgr daemon?  Why wouldn't we output the daemon name?  I guess it looks kind of funny because the daemon name is so often equal to the hostname, but it's still the name we use for most activities like looking at logs, doing a "ceph daemon..." CLI operation, etc.

Reply:
i think i was just blindly follow other implementations of `print_nodes()` without really thinking about the use case of this utility.

Reply:
Thanks for the all suggestion. I've updated the commit.
I can work on another pr to improve the MON/MDS output with their names in ‘node ls’